### PR TITLE
ch4/workq: create upper-layer workq request

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

[NOTE] this PR is based on #4694. The actual commits of this PR starts at `request: add MPIR_REQUEST_KIND__WORKQ`.

The current workq works by creating request at ch4-layer and then later pass down to shmmod/netmod layer at workq dispatch time. This is undesirable. For one, all netmod/shmmod implementation need deal with both the possibility of creating new request or using an existing request. When using an existing request, we have to always be careful that the same request may be concurrently progressed. This brings much complexity that is difficult to maintain. For two, creating requests at ch4-layer restricts the flexibility of netmod or shmmod deciding vci and all solutions will bring further complications.

In this PR, we implement the workq request as a pseudo request, very similar to how persistent request currently works. With pseudo request, netmod/shmmod layers will have full freedom in creating and progressing their own requests. The upper layer only need atomic access to the `cc_ptr` that is to be linked at workq dispatch time.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci --enable-ch4-mt=handoff --enable-izem-queue --with-zm-prefix=embedded
netmod: ch4:ofi

**Comment by hzhou** [comment]
@sagarth Do you have a chance to review this PR?

**Comment by sagarth** [comment]
@hzhou Sure, I'll have a look.

File: src/mpid/ch4/src/ch4i_workq_progress.c [hunk 1]
```diff
@@ -12,97 +12,110 @@ extern MPID_Thread_mutex_t MPIDI_workq_lock;
...
     int mpi_errno = MPI_SUCCESS;
-    MPIR_Request *req;
     MPI_Datatype datatype, origin_datatype, target_datatype;
 
     MPIR_Assert(workq_elemt != NULL);
 
     switch (workq_elemt->op) {
         case SEND:{
                 struct MPIDI_workq_send *wd = &workq_elemt->params.pt2pt.send;
-                req = wd->request;
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4i_workq_progress.c]: So I think my main concern is that this doubles the request allocation overhead of the workq path. That's something I know folks were keenly aware of when developing workq.

File: src/mpid/ch4/src/ch4_wait.h [hunk 2]
```diff
@@ -14,12 +14,44 @@ MPL_STATIC_INLINE_PREFIX int get_vci_wrapper(MPIR_Request * req)
     int vci;
     if (req->kind == MPIR_REQUEST_KIND__PREQUEST_RECV ||
         req->kind == MPIR_REQUEST_KIND__PREQUEST_SEND) {
-        if (req->u.persist.real_request) {
-            vci = MPIDI_Request_get_vci(req->u.persist.real_request);
+        MPIR_Request *real_req = req->u.persist.real_request;
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_wait.h:17]: I know this "real" request assumption is baked into the code right now, but if we could provide a way avoid it, we might be able to see some improvements in the persistent path. We'd need a way preset the VCI at persistent initialization time vs. when the communication is started.

This is orthogonal to this PR, but its something I've been thinking of for a while.

**Review by raffenet — COMMENTED** [review]
I think someone using workq is going to have to sign off on the double allocation change. That part sticks out to me.

File: src/mpid/ch4/src/ch4i_workq_progress.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4i_workq_progress.c]: I see. Let me think about how to cleanly (cleaner than current) handle pre-allocated requests.

**Comment by hzhou** [comment]
I guess there is always interest to avoid extra allocation of request objects from performance concerns. I'll work on the direction of how to handle these cases cleanly.
